adam-core/src/main/scala/org/bdgenomics/adam/models/SnpTable.scala

```diff
@@ -17,52 +17,111 @@
  */
 package org.bdgenomics.adam.models
 
-import org.bdgenomics.adam.rich.RichVariant
-import org.bdgenomics.adam.rich.DecadentRead._
-import org.bdgenomics.utils.misc.Logging
+import com.esotericsoftware.kryo.io.{ Input, Output }
+import com.esotericsoftware.kryo.{ Kryo, Serializer }
+import org.apache.spark.rdd.MetricsContext._
 import org.apache.spark.rdd.RDD
-import scala.collection.immutable._
-import scala.collection.mutable
+import org.bdgenomics.adam.instrumentation.Timers._
+import org.bdgenomics.adam.rdd.variant.VariantRDD
+import org.bdgenomics.adam.rich.DecadentRead.Residue
+import org.bdgenomics.utils.misc.Logging
+import scala.annotation.tailrec
+import scala.math.{ max, min }
 
 /**
  * A table containing all of the SNPs in a known variation dataset.
  *
  * @param table A map between a contig name and a set containing all coordinates
  *   where a point variant is known to exist.
  */
-class SnpTable(private val table: Map[String, Set[Long]]) extends Serializable with Logging {
-  log.info("SNP table has %s contigs and %s entries".format(
-    table.size,
-    table.values.map(_.size).sum))
+class SnpTable private[models] (
+    private[models] val indices: Map[String, (Int, Int)],
+    private[models] val sites: Array[Long]) extends Serializable with Logging {
 
-  /**
-   * Is there a known SNP at the reference location of this Residue?
-   */
-  def isMasked(residue: Residue): Boolean =
-    contains(residue.referencePosition)
+  private val midpoints: Map[String, Int] = {
+    @tailrec def pow2ceil(length: Int, i: Int = 1): Int = {
+      if (2 * i >= length) {
+        i
+      } else {
+        pow2ceil(length, 2 * i)
+      }
+    }
 
-  /**
-   * Is there a known SNP at the given reference location?
-   */
-  def contains(location: ReferencePosition): Boolean = {
-    val bucket = table.get(location.referenceName)
-    if (bucket.isEmpty) unknownContigWarning(location.referenceName)
-    bucket.exists(_.contains(location.pos))
+    indices.mapValues(p => {
+      val (start, end) = p
+      pow2ceil(end - start + 1)
+    })
   }
 
-  private val unknownContigs = new mutable.HashSet[String]
+  @tailrec private def binarySearch(rr: ReferenceRegion,
```

Comment:
The TL,DR is that since we're doing a set operation and not a full join, there's a significant performance reason to prefer this more specialized implementation.